### DISCUSSION TITLE
v8 causes imports and junit reporter to break

### DISCUSSION THREAD

**Discussion by ianhutson** [webdriverio/webdriverio#9576] [Q&A]
my team is trying to convert a typescript project to wdio v8, but we're running into quite a few issues. first, we can't run any tests until commenting out our junit reporter code in `wdio.conf`. it looks like this: 

<img width="660" alt="Screen Shot 2023-01-11 at 10 08 16 AM" src="https://user-images.githubusercontent.com/57458226/211841558-055a3661-706d-4e39-85e2-a01a52318309.png">

the error we get when running tests with junit reporter enabled is this:

<img width="820" alt="Screen Shot 2023-01-11 at 10 08 51 AM" src="https://user-images.githubusercontent.com/57458226/211844379-73c11493-ba8d-4c82-9a08-588707f57faa.png">

after commenting that out (ideally not a permanent solution), we get tsc build errors that look like this:

<img width="1319" alt="Screen Shot 2023-01-11 at 9 53 59 AM" src="https://user-images.githubusercontent.com/57458226/211838365-652838dc-d6d0-4baf-bf42-10b7301fd210.png">

following the error messages and through trial and error, it seems like i can convert evert filetype to `.mts`, which then changes all our import filepaths to `.mjs`, which fixes the ts issues. but we wrap quite a few exports in index files that look like this:

`export * from "./FileWithMultipleExportedTypesAndMethods1";`
`export * from "./FileWithMultipleExportedTypesAndMethods2";`
`export * from "./FileWithMultipleExportedTypesAndMethods3";`

and those wrapped modules seem to not work even when changing the file paths and extensions accordingly. rather than importing our modules per our old convention, `import { Xyz } from '@abc/xyz'`, it seems we have to drill into each module individually with the full file path, as well as make all of the extensions `.mts`. i tried deconstructing the imports per the error message's suggestion, but then i can't import types, or at least it isn't exactly clear to me how to do it.

the docs mention in the CommonJS to ESM transition section that `While this affected every single file it "should" be unrecognizable to you.` since this _is_ recognizable by us, does anyone have any tips for managing our modules in v8? we'd rather not ditch our export wrapper convention

**Comment by RossVertizan** [comment]
I spent a little while looking at this as I've been playing around with the whole CJS/ESM stuff recently.

On the first problem, it is a [known issue with the validator module](https://github.com/validatorjs/validator.js/issues/1759) and it appears that three is a [PR to fix it](https://github.com/validatorjs/validator.js/pull/1869) but it has not been integrated yet.  As a horrible hacky fix, the problem can be solved by adding a ```package.json``` file in the ```node_modules\validator\es``` directory with the content 
```json
{
    "type": "module"
}
```
and adding ```.js``` to every import statement in  ```node_modules\validator\es\index.js```
**I did say it was horrible and hacky, but it worked for me**

**Comment by christian-bromann** [comment]
duplicate of #9536

**Reply by ianhutson** [reply]
not exactly. half of my issue is discussed there, but there’s an entirely different issue after that

**Reply by christian-bromann** [reply]
The other half seems to be a issue with the TypeScript setup. Make sure it transpiles your test files into CJS if you have a CJS project. I can't say on top of my head how to fix the export/import issue but it should be doable.
